.teamcity/src/main/kotlin/model/FunctionalTestBucketProvider.kt

```diff
@@ -35,140 +35,50 @@ interface FunctionalTestBucketProvider {
-        val docs = allSubprojects.find { it.name == "docs" }!!
-        val docs1 = LargeSubprojectSplitBucket(docs, 1, true, listOf(TestClassTime("org.gradle.docs.samples.Bucket1SnippetsTest", "docsTest", -1)))
-        val docs2 = LargeSubprojectSplitBucket(docs, 2, true, listOf(TestClassTime("org.gradle.docs.samples.Bucket2SnippetsTest", "docsTest", -1)))
-        val docs3 = LargeSubprojectSplitBucket(docs, 3, true, listOf(TestClassTime("org.gradle.docs.samples.Bucket3SnippetsTest", "docsTest", -1)))
-        val docs4 = LargeSubprojectSplitBucket(
-            docs, 4, false, listOf(
-                TestClassTime("org.gradle.docs.samples.Bucket1SnippetsTest", "docsTest", -1),
-                TestClassTime("org.gradle.docs.samples.Bucket2SnippetsTest", "docsTest", -1),
-                TestClassTime("org.gradle.docs.samples.Bucket3SnippetsTest", "docsTest", -1)
-            )
-        )
-        return listOf(docs1, docs2, docs3, docs4)
+    override fun createFunctionalTestsFor(stage: Stage, testCoverage: TestCoverage): List<FunctionalTest> {
+        return buckets.mapIndexed { index, bucket -> bucket.createFunctionalTestsFor(model, stage, testCoverage, index + 1) }
     }
+}
 
-    private fun splitIntoBucketsExcludingSpecialBuckets(
-        specialSubprojectNames: List<String>,
-        validSubprojects: List<GradleSubproject>,
-        subProjectTestClassTimes: List<SubprojectTestClassTime>,
-        testCoverage: TestCoverage
-    ): List<BuildTypeBucket> {
-        val specialSubprojects = validSubprojects.filter { specialSubprojectNames.contains(it.name) }
-        val otherSubProjectTestClassTimes = subProjectTestClassTimes.filter { !specialSubprojectNames.contains(it.subProject.name) }
-        return splitIntoBuckets(
-            LinkedList(otherSubProjectTestClassTimes),
-            SubprojectTestClassTime::totalTime,
-            { largeElement: SubprojectTestClassTime, size: Int -> largeElement.split(size) },
-            { list: List<SubprojectTestClassTime> -> SmallSubprojectBucket(list) },
-            testCoverage.expectedBucketNumber - specialSubprojects.size,
-            MAX_PROJECT_NUMBER_IN_BUCKET
-        )
+class StatisticBasedFunctionalTestBucketProvider(val model: CIBuildModel, testBucketsJson: File) : FunctionalTestBucketProvider {
+    private val buckets: Map<TestCoverage, List<BuildTypeBucket>> by lazy {
+        val uuidToTestCoverage = model.stages.flatMap { it.functionalTests }.associateBy { it.uuid }
+        val testCoverageAndBuckets = JSON.parseArray(testBucketsJson.readText()) as JSONArray
+        testCoverageAndBuckets.map { testCoverageAndBucket ->
+            testCoverageAndBucket as JSONObject
+            val testCoverage: TestCoverage = uuidToTestCoverage.getValue(testCoverageAndBucket.getIntValue("testCoverageUuid"))
+            val buckets: List<BuildTypeBucket> = testCoverageAndBucket.getJSONArray("buckets").map {
```

Comment:
I there a way to parse this array with different types directly instead of doing this brittle check for some fields here? We should probably move the parsing to a Factory class, so we have this in one location close to the classes we serialize.

Reply:
Moved to a special method `fromJsonObject`.